feat(js): list api subpathStrategy parameter (#7794)
aws-amplify/docs: src/pages/[platform]/build-a-backend/storage/list-files/index.mdx

  };
}

You can list files without having to download all the files. You can do this by using the `list` API from the Amplify Library for Storage. You can also get properties individually for a file using the getProperties API.

## List Files

<InlineFilter filters={["react", "angular", "javascript", "vue", "nextjs", "react-native"]}>
```javascript
import { list } from 'aws-amplify/storage';

> try { const result = await list({ path: 'photos/', // Alternatively, path: ({identityId}) => `album/{identityId}/photos/` }); } catch (error) { console.log(error); }
> 
> ````

Reviewer: ````suggestion
const result = await list({
	path: 'photos/',
  // Alternatively, path: ({identityId}) => `album/{identityId}/photos/`
});
```
````

}
```
{/* in other files we're referring to paths instead of folders, can we be consistent on terminology? */}
Manually created folders will show up as files with a `size` of 0, but you can also match keys against a regex like `file.key.match(/\.[0-9a-z]+$/i)` to distinguish files from folders. Since "folders" are a virtual concept in Amazon S3, any file may declare any depth of folder just by having a `/` in its name. 

> You can explore the contents and subpaths of each "folder" by either requesting the contents of the entire path and parsing it, or by using the `subpathStrategy` option to limit your query to the path's file contents only.

Reviewer 2: ```suggestion
To access the contents and subpaths of a "folder", you have two options:

1. Request the entire path and parse the contents.
2. Use the subpathStrategy option to retrieve only the files within the specified path (i.e. exclude files under subpaths).
```

> ### Explore entire contents of a path

Reviewer 2: ```suggestion
### Get all nested files within a path
```

> If you need to list all the contents folders, you'll have to parse them accordingly to get an authoritative list of files and folders:

Reviewer 2: ```suggestion
This retrieves all files and folders under a given path. You may need to parse the result to get only the files within the specified path.
```

This places each item's data inside a special `__data` key.

### Exclude Subpaths

> When objects are organized using a prefix you can return only those objects within the `path` you've provided, and not those within nested subpaths.

Reviewer 3: this is a bit unclear. i think we need to explain the concept of subpath here. since we are newly introducing it.

Reviewer 4: Yeah basically a `subpath` is a common `path` between the main `path` and a `delimiter`.

For example, given the following keys in your `path` you may want to return only the jpg objects, and not the "vacation" subpath:

** `photos/photo1.jpg`
** `photos/photo2.jpg`
** `photos/vacation/`

This can be accomplished with the `subpathStrategy` option:

```js
const result = await list({ 
  path: "photos/",
  options:{
    subpathStrategy: { strategy:'exclude' }
  }
});

> ```

Reviewer 3: can we also include an example of the output. since this result does contain the `excludedSubPaths` as well?

Reviewer 4: It's possible to extract the object. e.g. `const =  { excludedSubpaths } = await list(..`

Reviewer 3: yes that would work.

The default delimiter character is '/', but this can be changed by supplying a custom delimiter:

```js
const result = await list({ 
  path: "photos-",
  options:{
    subpathStrategy: { 
      strategy:'exclude',
      delimiter: '-'
    }
  }
});
```

> 

Reviewer 2: Can we provide an example of the behavior for custom delimiter please?

This places each item's data inside a special `__data` key.

### Excluding subpaths

> In addition to using the `list` API to explore the complete contents of a path, you can also use it to explore the immediate contents of a path while excluding additional subpaths.

Reviewer 2: ```suggestion
In addition to using the `list` API to get all the contents of a path, you can also use it to get only the files within a path while excluding files under subpaths.
```

For example, given the following keys in your `path` you may want to return only the jpg object, and not the "vacation" subpath and its contents:

```
photos/photo1.jpg

> photos/vacation/`

Reviewer 3: ```suggestion
photos/vacation/
```

This places each item's data inside a special `__data` key.

### Excluding subpaths

In addition to using the `list` API to get all the contents of a path, you can also use it to get only the files within a path while excluding files under subpaths.

For example, given the following keys in your `path` you may want to return only the jpg object, and not the "vacation" subpath and its contents:

```
photos/photo1.jpg
photos/vacation/
```

This can be accomplished with the `subpathStrategy` option:

> ```js

Reviewer: can we also run this through prettier? the spacing and quote usage is mixed

const result = await list({ 
  path: "photos/",
  options:{
    subpathStrategy: { strategy:'exclude' }
  }
});
```

The response will include only the objects within the `photos/` path and will also communicate any excluded subpaths:

```js

> {

Reviewer: is this intended to show a json blob?

Author: Yes, following the pattern already used on line 54.

    excludedSubpaths: [
      'photos/vacation/'
    ],
    items: [
      {
        path: "photos/photo1.jpg",
        eTag: "30074401292215403a42b0739f3b5262",
        lastModified: "Thu Oct 08 2020 23:59:31 GMT+0800 (Singapore Standard Time)",
        size: 138256
      },
    ]
}
```

The default delimiter character is '/', but this can be changed by supplying a custom delimiter:

> ```js

Reviewer: same with this snippet
